build(gradle): upgrade gradle 6.4.2 >>> 8.2.1

DarkWeird opened the pull request:
### Contains

* Updating gradle from 6.4.2 -> 8.2.1
* Updating gradle's files for which became invalid

### How to test

* run `gradlew jar game` -> should build and run game
* run `gradlew unitTest` -> should run the unit test suites

### Remarks

Java 20 is not yet supported, see https://github.com/gradle/gradle/issues/23488.

Closes #5111

DarkWeird committed:
build(bumpup-gradle): upgrade gradle and code for it.

DarkWeird committed:
chore(bumpup-gradle): fix junitXmlReport.isEnabled -> required

skaldarnar commented on engine/build.gradle:
```diff
@@ -270,10 +270,6 @@ task copyResourcesToClasses(type: Copy) {
     mustRunAfter compileJava
 }
 
-tasks.named("classes") {
-    dependsOn(tasks.named("copyResourcesToClasses"))
-}
-
```
What happened to this? All the other changes look like simple drop-in replacements, but this is just removed. Do we just not need it anymore?

skaldarnar replied:
`gradlew unitTest` fails with a message seemingly related to this:

```
FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':engine-tests:copyResourcesToClasses' (type 'Copy').
  - Gradle detected a problem with the following location: '/home/jenkins/agent/workspace/Terasology_engine_PR-5109/engine-tests/build/classes'.
    
    Reason: Task ':engine-tests:compileTestJava' uses this output of task ':engine-tests:copyResourcesToClasses' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':engine-tests:copyResourcesToClasses' as an input of ':engine-tests:compileTestJava'.
      2. Declare an explicit dependency on ':engine-tests:copyResourcesToClasses' from ':engine-tests:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':engine-tests:copyResourcesToClasses' from ':engine-tests:compileTestJava' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/jenkins/agent/workspace/Terasology_engine_PR-5109/engine-tests/build/resources/main'.
    
    Reason: Task ':engine-tests:copyResourcesToClasses' uses this output of task ':engine-tests:processResources' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':engine-tests:processResources' as an input of ':engine-tests:copyResourcesToClasses'.
      2. Declare an explicit dependency on ':engine-tests:processResources' from ':engine-tests:copyResourcesToClasses' using Task#dependsOn.
      3. Declare an explicit dependency on ':engine-tests:processResources' from ':engine-tests:copyResourcesToClasses' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```

DarkWeird replied:
Hm. I just delete it , and game running.
And i see how fix this now.
Seems there we should declare second edge for task graph.

skaldarnar replied:
First I tried to add those missing dependencies, but it gets weird quickly, and you also have to add `DupliactesStrategy`s (I suspect because we wildly copy files around and don't care about duplicates or anything). I've added more and more `dependsOn` and `duplicatesStrategy = 'exclude'`, but in the end still failed to run the tests :pensive: 

My second attempt was to remove the `copyResourcesToClasses` task and pointers to it completely, from both `:engine` and `:engine-tests`. Eventually, the engine (tests) seemingly built, but then errors popped up when dealing with modules:

```
* What went wrong:
Some problems were found with the configuration of task ':modules:CoreWorlds:compileTestJava' (type 'JavaCompile').
  - Gradle detected a problem with the following location: '/home/skaldarnar/Code/movingblocks/ts-iota/modules/BiomesAPI/build/classes'.
    
    Reason: Task ':modules:CoreWorlds:compileTestJava' uses this output of task ':modules:BiomesAPI:syncModuleInfo' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':modules:BiomesAPI:syncModuleInfo' as an input of ':modules:CoreWorlds:compileTestJava'.
      2. Declare an explicit dependency on ':modules:BiomesAPI:syncModuleInfo' from ':modules:CoreWorlds:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':modules:BiomesAPI:syncModuleInfo' from ':modules:CoreWorlds:compileTestJava' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/skaldarnar/Code/movingblocks/ts-iota/modules/BiomesAPI/build/classes'.
    
    Reason: Task ':modules:CoreWorlds:compileTestJava' uses this output of task ':modules:BiomesAPI:syncAssets' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':modules:BiomesAPI:syncAssets' as an input of ':modules:CoreWorlds:compileTestJava'.
      2. Declare an explicit dependency on ':modules:BiomesAPI:syncAssets' from ':modules:CoreWorlds:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':modules:BiomesAPI:syncAssets' from ':modules:CoreWorlds:compileTestJava' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/skaldarnar/Code/movingblocks/ts-iota/modules/BiomesAPI/build/classes'.
    
    Reason: Task ':modules:CoreWorlds:compileTestJava' uses this output of task ':modules:BiomesAPI:syncDeltas' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':modules:BiomesAPI:syncDeltas' as an input of ':modules:CoreWorlds:compileTestJava'.
      2. Declare an explicit dependency on ':modules:BiomesAPI:syncDeltas' from ':modules:CoreWorlds:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':modules:BiomesAPI:syncDeltas' from ':modules:CoreWorlds:compileTestJava' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/skaldarnar/Code/movingblocks/ts-iota/modules/BiomesAPI/build/classes'.
    
    Reason: Task ':modules:CoreWorlds:compileTestJava' uses this output of task ':modules:BiomesAPI:syncOverrides' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':modules:BiomesAPI:syncOverrides' as an input of ':modules:CoreWorlds:compileTestJava'.
      2. Declare an explicit dependency on ':modules:BiomesAPI:syncOverrides' from ':modules:CoreWorlds:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':modules:BiomesAPI:syncOverrides' from ':modules:CoreWorlds:compileTestJava' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/skaldarnar/Code/movingblocks/ts-iota/modules/CoreAssets/build/classes'.
    
    Reason: Task ':modules:CoreWorlds:compileTestJava' uses this output of task ':modules:CoreAssets:syncModuleInfo' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':modules:CoreAssets:syncModuleInfo' as an input of ':modules:CoreWorlds:compileTestJava'.
      2. Declare an explicit dependency on ':modules:CoreAssets:syncModuleInfo' from ':modules:CoreWorlds:compileTestJava' using Task#dependsOn.
      3. Declare an explicit dependency on ':modules:CoreAssets:syncModuleInfo' from ':modules:CoreWorlds:compileTestJava' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.2/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.

```

skaldarnar replied:
Do we have to re-model all these dependencies to make Gradle happy? Is the new Gradle version just pointing us to places where the setup was "shaky" in the first place?

skaldarnar replied:
@BenjaminAmos @DarkWeird do you know what exactly is meant by

```
//TODO: Remove it  when gestalt will can to handle ProtectionDomain without classes (Resources)
```

DarkWeird replied:
> @BenjaminAmos @DarkWeird do you know what exactly is meant by
> 
> ```
> //TODO: Remove it  when gestalt will can to handle ProtectionDomain without classes (Resources)
> ```

Gestalt works with one source of module.(one jar, one dir etc)
Gradle split resouces and classes(compiled  classes) dirs when build sources.
More gradle creates separate classes dir on every sources (java/kotlin/groovy - etc)

skaldarnar reviewed (commented):
With **Java 11** compilation and starting the game works fine. :+1: 

With **Java 17** it compiles and I see the expected compilation errors.

With **Java 20** (Liberica) compilation fails with

```
❯ gradlew jar game                                                                                                                                                             (base) 
Processing facade facades:PC, including it as a sub-project
Processing facade facades:TeraEd, including it as a sub-project
> Task :build-logic:generateExternalPluginSpecBuilders UP-TO-DATE
> Task :build-logic:extractPrecompiledScriptPluginPlugins UP-TO-DATE
> Task :build-logic:compilePluginsBlocks UP-TO-DATE
> Task :build-logic:generatePrecompiledScriptPluginAccessors UP-TO-DATE
> Task :build-logic:generateScriptPluginAdapters UP-TO-DATE
> Task :build-logic:compileKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':build-logic:compileKotlin'.
> Error while evaluating property 'compilerOptions.jvmTarget' of task ':build-logic:compileKotlin'.
   > Failed to calculate the value of property 'jvmTarget'.
      > Unknown Kotlin JVM target: 20

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 531ms
6 actionable tasks: 1 executed, 5 up-to-date
```

DarkWeird committed:
build(bumpup-gradle): fix several declaration of `copyResourcesToClasses`. unitTests and integrationTests works now

BenjaminAmos committed:
build: fix undeclared implicit dependency gradle errors

BenjaminAmos committed:
build: fix more undeclared dependency gradle errors

jdrueckert commented:
green checkmarks, yey! :partying_face:

jdrueckert committed:
Merge branch 'develop' into build/bumpup-gradle-to-8.2

jdrueckert commented on subsystems/DiscordRPC/build.gradle.kts:
```diff
@@ -9,6 +9,12 @@ plugins {
 
 apply(from = "$rootDir/config/gradle/common.gradle")
 
+configure<SourceSetContainer> {
+    // Adjust output path (changed with the Gradle 6 upgrade, this puts it back)
+    main { java.destinationDirectory.set(File("$buildDir/classes")) }
+    test { java.destinationDirectory.set(File("$buildDir/testClasses")) }
+}
```
why do we need this now? how was it done before?

BenjaminAmos replied:
Gradle was less strict before, I think. The issue was that both the tests compilation and the main compilation were outputting to the `build/classes` directory and Gradle did not like that. This fixed the implicit dependency errors. I'm open to a better solution though, this was more of a quick-fix.

BenjaminAmos replied:
This snippet was originally from here:
https://github.com/MovingBlocks/Terasology/blob/324376a044db3e7e0395c0396de1ec3f0311a203/engine-tests/build.gradle#L46-L50.

jdrueckert replied:
I would argue it's reasonable to separate main compilation classes and test compilation classes, so I don't see a need to look for a better solution right now :woman_shrugging:

jdrueckert commented on subsystems/TypeHandlerLibrary/build.gradle.kts:
```diff
@@ -12,6 +12,12 @@ apply(from = "$rootDir/config/gradle/publish.gradle")
 group = "org.terasology.subsystems"
 version = project(":engine").version
 
+configure<SourceSetContainer> {
+    // Adjust output path (changed with the Gradle 6 upgrade, this puts it back)
+    main { java.destinationDirectory.set(File("$buildDir/classes")) }
+    test { java.destinationDirectory.set(File("$buildDir/testClasses")) }
+}
+
```
same question here

soloturn commented:
added a couple of lines in #5113 , which segvaults now with java-20, openjdk, on arch linux:
```
---------------  S U M M A R Y ------------

Command Line: -XX:MaxDirectMemorySize=512M -XX:+PrintCommandLineFlags -Xmx768M -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant org.terasology.engine.Terasology --homedir=.

Host: Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz, 4 cores, 15G, Arch Linux
Time: Mon Jul 10 06:26:14 2023 CEST elapsed time: 4.768798 seconds (0d 0h 0m 4s)

---------------  T H R E A D  ---------------

Current thread (0x00007fe948988640):  JavaThread "splashscreen-loop" daemon [_thread_in_native, id=114136, stack(0x00007fe92c7f8000,0x00007fe92c8f8000)]

Stack: [0x00007fe92c7f8000,0x00007fe92c8f8000],  sp=0x00007fe92c8f6320,  free space=1016k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [iris_dri.so+0xb560c]
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.lwjgl.opengl.GL11C.glClear(I)V+0
j  org.lwjgl.opengl.GL11.glClear(I)V+1
j  org.terasology.splash.glfw.graphics.Renderer.clear()V+3
j  org.terasology.engine.GLFWSplashScreen.run()V+457
j  java.lang.Thread.runWith(Ljava/lang/Object;Ljava/lang/Runnable;)V+5 java.base@20.0.1
j  java.lang.Thread.run()V+19 java.base@20.0.1
v  ~StubRoutines::call_stub 0x00007fe937d37cc6

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x00007feeb60ba9f0

```

jdrueckert commented:
In https://github.com/MovingBlocks/gestalt/pull/136 there's additional changes to the gradle wrapper... we might want to consider adopting those changes here, too. Likely they are automatically generated/updated by gradle - I just don't know how to trigger that thinking Using gradlew after updating the version in gradlew-wrapper.properties didn't...

BenjaminAmos committed:
build: update groovyw wrapper to work with Gradle 8

BenjaminAmos commented:
The gradle upgrade broke the `groovyw` command, so I've replaced the groovyw wrapper a newer one from [`2.1.0`](https://github.com/MovingBlocks/groovy-wrapper/releases/tag/v2.1.0) to fix it.

skaldarnar committed:
Merge remote-tracking branch 'origin/develop' into build/bumpup-gradle-to-8.2

# Conflicts:
#	facades/TeraEd/build.gradle

soloturn commented:
#5113 does not crash java-20 jvm any more when as well merging #5107 . java-11 still works. imo these 2 plus #5109 could be merged now.

DarkWeird committed:
Merge branch 'develop' into build/bumpup-gradle-to-8.2

soloturn commented:
@DarkWeird you mind merging or cherry-picking https://github.com/MovingBlocks/Terasology/pull/5113/commits/0777f4f1223e8c23c35d83fc26ce1212a89e94d3 , #5113? or you prefer this one to be targetted to develop later?

soloturn committed:
use kotlin-1.9, force java-11 target. (#5113)

kotlin-1.9 knows java-20. setting java-11 as target a little more harsh
than currently.

DarkWeird committed:
Merge branch 'develop' into build/bumpup-gradle-to-8.2

soloturn commented:
in arch linux, with java-20, jvm does not crash any more using this branch. the error is the same as mentioned by @BenjaminAmos in #3976. so the branch could be merged as is.

```
10:22:52.761 [main] ERROR o.t.engine.core.modes.StateLoading - Error while loading org.terasology.engine.core.modes.loadProcesses.AwaitCharacterSpawn@28200d43
java.lang.IllegalArgumentException: Cannot access field java.util.concurrent.ExecutorCompletionService$QueueingFuture.task
        at org.terasology.engine.utilities.ReflectionUtil.readField(ReflectionUtil.java:497)
        at org.terasology.engine.world.chunks.pipeline.ChunkProcessingPipeline.lambda$unwrappingComporator$0(ChunkProcessingPipeline.java:93)
        at java.base/java.util.concurrent.PriorityBlockingQueue.siftUpUsingComparator(PriorityBlockingQueue.java:366)
        at java.base/java.util.concurrent.PriorityBlockingQueue.offer(PriorityBlockingQueue.java:477)
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1368)
        at java.base/java.util.concurrent.ExecutorCompletionService.submit(ExecutorCompletionService.java:186)
        at org.terasology.engine.world.chunks.pipeline.ChunkProcessingPipeline.invokeGeneratorTask(ChunkProcessingPipeline.java:239)
        at org.terasology.engine.world.chunks.localChunkProvider.LocalChunkProvider.createOrLoadChunk(LocalChunkProvider.java:120)
        at org.terasology.engine.world.chunks.localChunkProvider.RelevanceSystem.updateRelevance(RelevanceSystem.java:135)
        at org.terasology.engine.world.chunks.localChunkProvider.RelevanceSystem.update(RelevanceSystem.java:223)
        at org.terasology.engine.core.modes.loadProcesses.AwaitCharacterSpawn.step(AwaitCharacterSpawn.java:37)
        at org.terasology.engine.core.modes.StateLoading.update(StateLoading.java:259)
        at org.terasology.engine.core.TerasologyEngine.tick(TerasologyEngine.java:512)
        at org.terasology.engine.core.TerasologyEngine.mainLoop(TerasologyEngine.java:472)
        at org.terasology.engine.core.TerasologyEngine.runMain(TerasologyEngine.java:448)
        at org.terasology.engine.core.TerasologyEngine.run(TerasologyEngine.java:414)
        at org.terasology.engine.Terasology.call(Terasology.java:189)
        at org.terasology.engine.Terasology.call(Terasology.java:69)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1933)
        at picocli.CommandLine.access$1200(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2326)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2291)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)
        at picocli.CommandLine.execute(CommandLine.java:2058)
        at org.terasology.engine.Terasology.main(Terasology.java:138)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.concurrent.Future java.util.concurrent.ExecutorCompletionService$QueueingFuture.task accessible: module java.base does not "opens java.util.concurrent" to unnamed module @1684264
        at java.base/java.lang.reflect.AccessibleObject.throwInaccessibleObjectException(AccessibleObject.java:387)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:363)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:311)
        at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:181)
        at java.base/java.lang.reflect.Field.setAccessible(Field.java:175)
        at org.terasology.engine.utilities.ReflectionUtil.readField(ReflectionUtil.java:491)
        ... 25 common frames omitted
```

soloturn commented:
found another one, sputbug needs to be pudated: #5126. if you could merge @DarkWeird ?

BenjaminAmos commented:
Gradle `8.3-rc2` [now supports Java 20](https://docs.gradle.org/8.3-rc-2/release-notes.html#full-support-for-java-20). I don't believe that we're aiming for Java 20 support yet but it should now be possible to test against it with a slight gradle upgrade.

jdrueckert committed:
build: bump gradle 8.2 -> 8.2.1

jdrueckert committed:
Merge branch 'develop' into build/bumpup-gradle-to-8.2

jdrueckert committed:
chore: update gradle wrapper

jdrueckert commented:
@DarkWeird gradlew output mentions the following:
```
This version of Gradle expects version '4.0.14' of the `kotlin-dsl` plugin but version '4.1.0' has been applied to project ':build-logic'. Let Gradle control the version of `kotlin-dsl` by removing any explicit `kotlin-dsl` version constraints from your build logic.
WARNING: Unsupported Kotlin plugin version.
The `embedded-kotlin` and `kotlin-dsl` plugins rely on features of Kotlin `1.8.20` that might work differently than in the requested version `1.9.0`.
```

Guess we might want to fix that as part of this PR?

jdrueckert commented:
Also, there's a bunch of warnings which I think we don't need to fix in this PR, though.
I created an issue to track them: #5127

jdrueckert committed:
Merge branch 'develop' into build/bumpup-gradle-to-8.2

soloturn committed:
ci: update spotbugs version for gradle 7+, java-17, java-20 (#5126)

jdrueckert committed:
Merge branch 'develop' into build/bumpup-gradle-to-8.2
